File: Restoring trees.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long int ll;
#define fast ios_base::sync_with_stdio(0),cin.tie(0),cout.tie(0);

int n;
int ans[300005];
int start[300005];
int finish[300005];
int timer;

void dfs(int parent)
{
    int node = start[timer];
    
    ans[node] = parent;
    
    timer++;
    
    while(finish[node] > timer)
        dfs(node);
}

int main()
{
    fast
    
    cin >> n;
    
    for(int i = 1; i <= n; i++)
    {
        int x;
        
        cin >> x;
        
        start[x] = i;
    }
    
    for(int i = 1; i <= n; i++)
        cin >> finish[i];
        
    timer = 0;
    dfs(0);
    
    for(int i = 1; i <= n; i++)
        cout << ans[i] << ' ';
}

File: Restoring Trees.java
import java.util.*;
import java.io.*;

class TestClass {
    
    static int start[],finish[],ans[],timer,n;
    
    static void dfs(int parent)
    {
        int node = start[timer];
        ans[node] = parent;
        
        timer++;
        
        while(finish[node] > timer)
            dfs(node);
    }
    public static void main(String args[]) throws Exception {
        
        Scanner in = new Scanner(System.in);
        PrintWriter out = new PrintWriter(System.out);
        
        n = in.nextInt();
        
        start = new int[n+1];
        finish = new int[n+1];
        ans = new int[n+1];
        
        for(int i = 1; i <= n; i++)
        {
            int temp = in.nextInt();
            start[temp] = i;
        }
        
        for(int i = 1; i <= n; i++)
            finish[i] = in.nextInt();
        
        timer = 0;
        dfs(0);
        
        for(int i = 1; i <= n; i++)
            out.print(ans[i] + " ");
            
        out.close();
    }
}
